flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/SpeculativeExecutionVertex.java

```diff
@@ -295,6 +296,14 @@ void cachePartitionInfo(PartitionInfo partitionInfo) {
                 "Method is not supported in SpeculativeExecutionVertex.");
     }
 
+    @Override
+    protected boolean needMarkPartitionFinished(ResultPartitionType resultPartitionType) {
+        // for speculative execution, only blocking or hybrid full result partition need mark
+        // finished.
+        return resultPartitionType.isBlockingOrBlockingPersistentResultPartition()
+                || resultPartitionType == ResultPartitionType.HYBRID_FULL;
+    }
```

Comment:
I think this is not only for the speculative execution vertex. By default, in speculative execution we want the downstream to only consume the finished upstream partitions, and in non-speculative execution we want the downstream to also consume the unfinished partitions. However, user should also be able to choose other behaviors, consuming unfinished partitions even in speculative mode, or consuming only finished partitions in non-speculative mode.

Reply:
Consuming unfinished partitions in speculative mode may be difficult to support at present, but consuming only finished partitions in non-speculative mode sounds good.